### PR TITLE
Fix _empty_per_channel_affine_quantized to be less hacky

### PR DESCRIPTION
**Pull request by dzhulgakov**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26676 Allow per-channel QTensor accept any floating type for scales
* #26675 Per-channel quantized tensor to have only a single axis
* **#26243 Fix _empty_per_channel_affine_quantized to be less hacky**

This is an attempt to fix _empty_per_channel_affine_quantized to be more sane. It's a factory function that nevertheless receives a Tensor argument and it throws the codegen off course.

Before people did a hacky workaround of appending _like to the function name to trick codegen, it also required non-natural argument order.

This PR explicitly allows to override the 'category' of the function to make codegen do the right thing. Now name and the argument order (in C++) make more sense.

Gist with changes to codegen: https://gist.github.com/dzhulgakov/87050fcf7527730bea6483600f8c4897 (from ./tools/git_add_generated_dirs.sh)

Differential Revision: [D17443221](https://our.internmc.facebook.com/intern/diff/D17443221)

### CODE REVIEW

File: aten/src/ATen/native/README.md [hunk 1]
```diff
@@ -99,7 +99,9 @@ signature.
 Functions with no tensor inputs are called *factory functions*, and
 are handled specially by code generation.  If your function is behaving
 differently than another example, check first and see if one is a
-factory while another is not.
+factory while another is not. In some rare cases, factory function might have a
+tensor argument. In this case it'd be marked with 'category_override: factory'
```
Comment (jamesr66a) [inline aten/src/ATen/native/README.md]: `it'd be marked` -> `mark it`

**Comment by jamesr66a** [comment]
Can you post a diff of the codegen'ed files before and after this change?

File: aten/src/ATen/native/quantized/QTensor.cpp [hunk 2]
```diff
@@ -142,10 +142,10 @@ Tensor per_channel_affine_qtensor_cpu(
     const Tensor& scales,
     const Tensor& zero_points,
     IntArrayRef axis) {
-  Tensor dst = at::_empty_per_channel_affine_quantized_like(
+  Tensor dst = at::_empty_per_channel_affine_quantized(
```
Comment (jerryzh168) [inline aten/src/ATen/native/quantized/QTensor.cpp]: what should be the order of arguments?
should we put size in the end so it is closer to options?

**Review by raghuramank100 — COMMENTED** [review]
Looks good to me, will let folks more familiar with this part of the code approve.

**Review by gchanan — COMMENTED** [review]
ya I'd like to see the codegen changes.

I think the original issue is we didn't know how to dispatch these things.  Because we usually dispatch based on the tensor argument but in this case, I'm not even sure what we were dispatching on.  I also think @ezyang's recent multi dispatch changes might have affected this, because I know he was changing how "new" functions worked.

File: aten/src/ATen/native/quantized/QTensor.cpp (hunk 2 above)
Comment (dzhulgakov) [reply aten/src/ATen/native/quantized/QTensor.cpp]: I think sizes should be first - similarly to how we do it for at::full or at::rand

**Comment by dzhulgakov** [comment]
Gist with changes to codegen: https://gist.github.com/dzhulgakov/87050fcf7527730bea6483600f8c4897 (from ./tools/git_add_generated_dirs.sh)

**Comment by ezyang** [comment]
The conflict with my changes looks very mild and I can help resolve the changes if I land before this one. (Or if you land before me, I can fix it in my patch, but I'm landing right now and hope it will work this time ;)

File: aten/src/ATen/native/native_functions.yaml [hunk 3]
```diff
@@ -991,7 +991,9 @@
   dispatch:
     QuantizedCPU: empty_affine_quantized_cpu
 
-- func: _empty_per_channel_affine_quantized_like(Tensor self, Tensor zero_points, int[] size, int[] axis, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None, MemoryFormat? memory_format=contiguous_format) -> Tensor
+# it's a factory function receiving a tensor argument, thus overriding explicitly
+- func: _empty_per_channel_affine_quantized(int[] size, *, Tensor scales, Tensor zero_points, int[] axis, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None, MemoryFormat? memory_format=contiguous_format) -> Tensor
```
Comment (gchanan) [inline aten/src/ATen/native/native_functions.yaml]: I think this is all a pre-existing issue with TensorOptions, but I don't see how the dispatch makes sense here.

For example, if I don't specify a dtype, I won't try to dispatch to QuantizedCPU and presumably I'll just get some crappy error message about _empty_per_channel_affine_quantized not being implemented for CPU backend.

But really...dtype is not optional.  So maybe we don't want to do anything until @izdeby breaks apart TensorOptions, but it seems reasonable to provide a CPU overload and to give a real error message here.

**Review by gchanan — APPROVED** [review]

File: aten/src/ATen/native/native_functions.yaml (hunk 3 above)
Comment (dzhulgakov) [reply aten/src/ATen/native/native_functions.yaml]: Good idea, let me fix the error message.Btw, it applies to other qtensor factory functions too

**Pull request closed without merging**